File: sonet-client/android/app/src/main/java/xyz/sonet/app/viewmodels/NotificationsViewModel.kt
package xyz.sonet.app.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import xyz.sonet.app.grpc.SonetGRPCClient
import xyz.sonet.app.grpc.SonetConfiguration
import xyz.sonet.app.grpc.proto.*
import java.util.*

class NotificationsViewModel(application: Application) : AndroidViewModel(application) {
    
    // MARK: - State Flows
    private val _notifications = MutableStateFlow<List<NotificationItem>>(emptyList())
    val notifications: StateFlow<List<NotificationItem>> = _notifications.asStateFlow()
    
    private val _filteredNotifications = MutableStateFlow<List<NotificationItem>>(emptyList())
    val filteredNotifications: StateFlow<List<NotificationItem>> = _filteredNotifications.asStateFlow()
    
    private val _selectedFilter = MutableStateFlow(NotificationFilter.ALL)
    val selectedFilter: StateFlow<NotificationFilter> = _selectedFilter.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    
    private val _unreadCount = MutableStateFlow(0)
    val unreadCount: StateFlow<Int> = _unreadCount.asStateFlow()
    
    private val _showAppUpdate = MutableStateFlow(false)
    val showAppUpdate: StateFlow<Boolean> = _showAppUpdate.asStateFlow()
    
    private val _appUpdateInfo = MutableStateFlow<AppUpdateInfo?>(null)
    val appUpdateInfo: StateFlow<AppUpdateInfo?> = _appUpdateInfo.asStateFlow()
    
    private val _notificationPreferences = MutableStateFlow(NotificationPreferences())
    val notificationPreferences: StateFlow<NotificationPreferences> = _notificationPreferences.asStateFlow()
    
    // MARK: - Private Properties
    private val grpcClient = SonetGRPCClient(application, SonetConfiguration.development)
    private var notificationStream: kotlinx.coroutines.Job? = null
    private var refreshJob: kotlinx.coroutines.Job? = null
    
    // MARK: - Notification Filters
    enum class NotificationFilter(val title: String, val icon: String) {
        ALL("All", "bell"),
        MENTIONS("Mentions", "at"),
        LIKES("Likes", "heart"),
        REPOSTS("Reposts", "arrow_2_squarepath"),
        FOLLOWS("Follows", "person_badge_plus"),
        REPLIES("Replies", "bubble_left"),
        HASHTAGS("Hashtags", "number"),
        APP_UPDATES("App Updates", "app_badge")
    }
    
    // MARK: - Initialization
    init {
        setupNotificationStream()
        setupRefreshTimer()
        loadNotificationPreferences()
        requestNotificationPermissions()
    }
    
    // MARK: - Public Methods
    fun loadNotifications() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            
            try {
                val response = grpcClient.getNotifications(0, 50)
                _notifications.value = response.notificationsList.map { NotificationItem.from(it) }
                applyFilter()
                updateUnreadCount()
            } catch (error: Exception) {
                _error.value = "Failed to load notifications: ${error.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun refreshNotifications() {
        loadNotifications()
    }
    
    fun selectFilter(filter: NotificationFilter) {
        _selectedFilter.value = filter
        applyFilter()
    }
    
    fun markAsRead(notification: NotificationItem) {
        if (notification.isRead) return
        
        viewModelScope.launch {
            try {
                val response = grpcClient.markNotificationAsRead(notification.id)
                if (response.success) {
                    val currentNotifications = _notifications.value.toMutableList()
                    val index = currentNotifications.indexOfFirst { it.id == notification.id }
                    if (index != -1) {
                        currentNotifications[index] = notification.copy(isRead = true)
                        _notifications.value = currentNotifications
                        updateUnreadCount()
                    }
                }
            } catch (error: Exception) {
                // Handle error silently
            }
        }
    }
    
    fun markAllAsRead() {
        viewModelScope.launch {
            try {
                val response = grpcClient.markAllNotificationsAsRead()
                if (response.success) {
                    val currentNotifications = _notifications.value.map { it.copy(isRead = true) }
                    _notifications.value = currentNotifications
                    updateUnreadCount()
                }
            } catch (error: Exception) {
                // Handle error silently
            }
        }
    }
    
    fun deleteNotification(notification: NotificationItem) {
        viewModelScope.launch {
            try {
                val response = grpcClient.deleteNotification(notification.id)
                if (response.success) {
                    val currentNotifications = _notifications.value.toMutableList()
                    currentNotifications.removeAll { it.id == notification.id }
                    _notifications.value = currentNotifications
                    applyFilter()
                    updateUnreadCount()
                }
            } catch (error: Exception) {
                // Handle error silently
            }
        }
    }
    
    fun clearAllNotifications() {
        viewModelScope.launch {
            try {
                val response = grpcClient.clearAllNotifications()
                if (response.success) {
                    _notifications.value = emptyList()
                    applyFilter()
                    updateUnreadCount()
                }
            } catch (error: Exception) {
                // Handle error silently
            }
        }
    }
    
    fun checkForAppUpdates() {
        viewModelScope.launch {
            try {
                val response = grpcClient.checkForAppUpdates()
                if (response.hasUpdate) {
                    _appUpdateInfo.value = AppUpdateInfo.from(response)
                    _showAppUpdate.value = true
                }
            } catch (error: Exception) {
                // Handle error silently
            }
        }
    }
    
    fun updateNotificationPreferences() {
        viewModelScope.launch {
            try {
                val response = grpcClient.updateNotificationPreferences(_notificationPreferences.value.toGRPC())
                if (response.success) {
                    // Preferences updated successfully
                }
            } catch (error: Exception) {
                // Handle error silently
            }
        }
    }
    
    // MARK: - Private Methods
    private fun setupNotificationStream() {
        notificationStream = viewModelScope.launch {
            try {
                grpcClient.notificationStream().collect { notification ->
                    handleNewNotification(notification)
                }
            } catch (error: Exception) {
                _error.value = "Notification stream error: ${error.message}"
            }
        }
    }
    
    private fun setupRefreshTimer() {
        refreshJob = viewModelScope.launch {
            kotlinx.coroutines.delay(30000) // 30 seconds
            refreshNotifications()
        }
    }
    
    private fun handleNewNotification(notification: xyz.sonet.app.grpc.proto.Notification) {
        val newNotification = NotificationItem.from(notification)
        
        // Add to beginning of list
        val currentNotifications = _notifications.value.toMutableList()
        currentNotifications.add(0, newNotification)
        _notifications.value = currentNotifications
        
        // Apply current filter
        applyFilter()
        
        // Update unread count
        updateUnreadCount()
        
        // Show local notification if app is in background
        // This would be implemented with WorkManager or AlarmManager
    }
    
    private fun applyFilter() {
        val currentNotifications = _notifications.value
        
        _filteredNotifications.value = when (_selectedFilter.value) {
            NotificationFilter.ALL -> currentNotifications
            NotificationFilter.MENTIONS -> currentNotifications.filter { it.type == NotificationType.MENTION }
            NotificationFilter.LIKES -> currentNotifications.filter { it.type == NotificationType.LIKE }
            NotificationFilter.REPOSTS -> currentNotifications.filter { it.type == NotificationType.REPOST }
            NotificationFilter.FOLLOWS -> currentNotifications.filter { it.type == NotificationType.FOLLOW }
            NotificationFilter.REPLIES -> currentNotifications.filter { it.type == NotificationType.REPLY }
            NotificationFilter.HASHTAGS -> currentNotifications.filter { it.type == NotificationType.HASHTAG }
            NotificationFilter.APP_UPDATES -> currentNotifications.filter { it.type == NotificationType.APP_UPDATE }
        }
    }
    
    private fun updateUnreadCount() {
        _unreadCount.value = _notifications.value.count { !it.isRead }
    }
    
    private fun loadNotificationPreferences() {
        // Load from SharedPreferences
        // This would be implemented for preference persistence
    }
    
    private fun requestNotificationPermissions() {
        // Request notification permissions
        // This would be implemented with the notification permission request
    }
    
    override fun onCleared() {
        super.onCleared()
        notificationStream?.cancel()
        refreshJob?.cancel()
    }
}

// MARK: - Data Models
data class NotificationItem(
    val id: String,
    val type: NotificationType,
    val title: String,
    val message: String,
    val timestamp: Long,
    val isRead: Boolean,
    val sender: UserProfile?,
    val targetNote: Note?,
    val targetUser: UserProfile?,
    val metadata: Map<String, String>
) {
    companion object {
        fun from(grpcNotification: xyz.sonet.app.grpc.proto.Notification): NotificationItem {
            return NotificationItem(
                id = grpcNotification.notificationId,
                type = NotificationType.from(grpcNotification.type),
                title = grpcNotification.title,
                message = grpcNotification.message,
                timestamp = grpcNotification.timestamp.seconds * 1000,
                isRead = grpcNotification.isRead,
                sender = if (grpcNotification.hasSender) UserProfile.from(grpcNotification.sender) else null,
                targetNote = if (grpcNotification.hasTargetNote) Note.from(grpcNotification.targetNote) else null,
                targetUser = if (grpcNotification.hasTargetUser) UserProfile.from(grpcNotification.targetUser) else null,
                metadata = grpcNotification.metadataMap
            )
        }
    }
}

enum class NotificationType(val icon: String) {
    LIKE("heart_fill"),
    REPLY("bubble_left_fill"),
    REPOST("arrow_2_squarepath_fill"),
    FOLLOW("person_badge_plus_fill"),
    MENTION("at"),
    HASHTAG("number"),
    APP_UPDATE("app_badge_fill"),
    SYSTEM("bell_fill");
    
    companion object {
        fun from(grpcType: xyz.sonet.app.grpc.proto.NotificationType): NotificationType {
            return when (grpcType) {
                xyz.sonet.app.grpc.proto.NotificationType.NOTIFICATION_TYPE_LIKE -> LIKE
                xyz.sonet.app.grpc.proto.NotificationType.NOTIFICATION_TYPE_REPLY -> REPLY
                xyz.sonet.app.grpc.proto.NotificationType.NOTIFICATION_TYPE_REPOST -> REPOST
                xyz.sonet.app.grpc.proto.NotificationType.NOTIFICATION_TYPE_FOLLOW -> FOLLOW
                xyz.sonet.app.grpc.proto.NotificationType.NOTIFICATION_TYPE_MENTION -> MENTION
                xyz.sonet.app.grpc.proto.NotificationType.NOTIFICATION_TYPE_HASHTAG -> HASHTAG
                xyz.sonet.app.grpc.proto.NotificationType.NOTIFICATION_TYPE_APP_UPDATE -> APP_UPDATE
                xyz.sonet.app.grpc.proto.NotificationType.NOTIFICATION_TYPE_SYSTEM -> SYSTEM
                else -> SYSTEM
            }
        }
    }
}

data class AppUpdateInfo(
    val id: String,
    val version: String,
    val title: String,
    val description: String,
    val changelog: List<String>,
    val isRequired: Boolean,
    val downloadUrl: String,
    val releaseDate: Long
) {
    companion object {
        fun from(grpcUpdate: xyz.sonet.app.grpc.proto.AppUpdate): AppUpdateInfo {
            return AppUpdateInfo(
                id = grpcUpdate.updateId,
                version = grpcUpdate.version,
                title = grpcUpdate.title,
                description = grpcUpdate.description,
                changelog = grpcUpdate.changelogList,
                isRequired = grpcUpdate.isRequired,
                downloadUrl = grpcUpdate.downloadUrl,
                releaseDate = grpcUpdate.releaseDate.seconds * 1000
            )
        }
    }
}

data class NotificationPreferences(
    val pushEnabled: Boolean = true,
    val emailEnabled: Boolean = true,
    val likeNotifications: Boolean = true,
    val replyNotifications: Boolean = true,
    val repostNotifications: Boolean = true,
    val followNotifications: Boolean = true,
    val mentionNotifications: Boolean = true,
    val hashtagNotifications: Boolean = true,
    val appUpdateNotifications: Boolean = true,
    val systemNotifications: Boolean = true,
    val quietHoursEnabled: Boolean = false,
    val quietHoursStart: Long = 0L,
    val quietHoursEnd: Long = 0L
) {
    fun toGRPC(): xyz.sonet.app.grpc.proto.NotificationPreferences {
        return xyz.sonet.app.grpc.proto.NotificationPreferences.newBuilder()
            .setPushEnabled(pushEnabled)
            .setEmailEnabled(emailEnabled)
            .setLikeNotifications(likeNotifications)
            .setReplyNotifications(replyNotifications)
            .setRepostNotifications(repostNotifications)
            .setFollowNotifications(followNotifications)
            .setMentionNotifications(mentionNotifications)
            .setHashtagNotifications(hashtagNotifications)
            .setAppUpdateNotifications(appUpdateNotifications)
            .setSystemNotifications(systemNotifications)
            .setQuietHoursEnabled(quietHoursEnabled)
            .setQuietHoursStart(Timestamp.newBuilder().setSeconds(quietHoursStart / 1000).build())
            .setQuietHoursEnd(Timestamp.newBuilder().setSeconds(quietHoursEnd / 1000).build())
            .build()
    }
}

File: sonet-ios/Sonet/Views/Notifications/NotificationsViewModel.swift
import Foundation
import Combine
import SwiftUI
import UserNotifications

@MainActor
class NotificationsViewModel: ObservableObject {
    
    // MARK: - Published Properties
    @Published var notifications: [NotificationItem] = []
    @Published var filteredNotifications: [NotificationItem] = []
    @Published var selectedFilter: NotificationFilter = .all
    @Published var isLoading = false
    @Published var error: String?
    @Published var unreadCount = 0
    @Published var showAppUpdate = false
    @Published var appUpdateInfo: AppUpdateInfo?
    @Published var notificationPreferences: NotificationPreferences = NotificationPreferences()
    
    // MARK: - Private Properties
    private let grpcClient: SonetGRPCClient
    private var cancellables = Set<AnyCancellable>()
    private var notificationStream: AnyCancellable?
    private var refreshTimer: Timer?
    
    // MARK: - Initialization
    init(grpcClient: SonetGRPCClient) {
        self.grpcClient = grpcClient
        setupNotificationStream()
        setupRefreshTimer()
        loadNotificationPreferences()
        requestNotificationPermissions()
    }
    
    // MARK: - Notification Filters
    enum NotificationFilter: String, CaseIterable {
        case all = "All"
        case mentions = "Mentions"
        case likes = "Likes"
        case reposts = "Reposts"
        case follows = "Follows"
        case replies = "Replies"
        case hashtags = "Hashtags"
        case appUpdates = "App Updates"
        
        var icon: String {
            switch self {
            case .all: return "bell"
            case .mentions: return "at"
            case .likes: return "heart"
            case .reposts: return "arrow.2.squarepath"
            case .follows: return "person.badge.plus"
            case .replies: return "bubble.left"
            case .hashtags: return "number"
            case .appUpdates: return "app.badge"
            }
        }
        
        var color: Color { .primary }
    }
    
    // MARK: - Public Methods
    func loadNotifications() {
        Task {
            isLoading = true
            error = nil
            
            do {
                let response = try await grpcClient.getNotifications(page: 0, pageSize: 50)
                await MainActor.run {
                    self.notifications = response.notifications.map { NotificationItem(from: $0) }
                    self.applyFilter()
                    self.updateUnreadCount()
                }
            } catch {
                await MainActor.run {
                    self.error = "Failed to load notifications: \(error.localizedDescription)"
                }
            }
            
            await MainActor.run {
                self.isLoading = false
            }
        }
    }
    
    func refreshNotifications() {
        loadNotifications()
    }
    
    func selectFilter(_ filter: NotificationFilter) {
        selectedFilter = filter
        applyFilter()
    }
    
    func markAsRead(_ notification: NotificationItem) {
        guard !notification.isRead else { return }
        
        Task {
            do {
                let response = try await grpcClient.markNotificationAsRead(notificationId: notification.id)
                if response.success {
                    await MainActor.run {
                        if let index = self.notifications.firstIndex(where: { $0.id == notification.id }) {
                            self.notifications[index].isRead = true
                        }
                        self.updateUnreadCount()
                    }
                }
            } catch {
                // Handle error silently
            }
        }
    }
    
    func markAllAsRead() {
        Task {
            do {
                let response = try await grpcClient.markAllNotificationsAsRead()
                if response.success {
                    await MainActor.run {
                        for index in self.notifications.indices {
                            self.notifications[index].isRead = true
                        }
                        self.updateUnreadCount()
                    }
                }
            } catch {
                // Handle error silently
            }
        }
    }
    
    func deleteNotification(_ notification: NotificationItem) {
        Task {
            do {
                let response = try await grpcClient.deleteNotification(notificationId: notification.id)
                if response.success {
                    await MainActor.run {
                        self.notifications.removeAll { $0.id == notification.id }
                        self.applyFilter()
                        self.updateUnreadCount()
                    }
                }
            } catch {
                // Handle error silently
            }
        }
    }
    
    func clearAllNotifications() {
        Task {
            do {
                let response = try await grpcClient.clearAllNotifications()
                if response.success {
                    await MainActor.run {
                        self.notifications.removeAll()
                        self.applyFilter()
                        self.updateUnreadCount()
                    }
                }
            } catch {
                // Handle error silently
            }
        }
    }
    
    func checkForAppUpdates() {
        Task {
            do {
                let response = try await grpcClient.checkForAppUpdates()
                if response.hasUpdate {
                    await MainActor.run {
                        self.appUpdateInfo = AppUpdateInfo(from: response)
                        self.showAppUpdate = true
                    }
                }
            } catch {
                // Handle error silently
            }
        }
    }
    
    func updateNotificationPreferences() {
        Task {
            do {
                let response = try await grpcClient.updateNotificationPreferences(preferences: notificationPreferences.toGRPC())
                if response.success {
                    // Preferences updated successfully
                }
            } catch {
                // Handle error silently
            }
        }
    }
    
    // MARK: - Private Methods
    private func setupNotificationStream() {
        // Set up real-time notification stream
        notificationStream = grpcClient.notificationStream()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { completion in
                    switch completion {
                    case .finished:
                        break
                    case .failure(let error):
                        self.error = "Notification stream error: \(error.localizedDescription)"
                    }
                },
                receiveValue: { notification in
                    self.handleNewNotification(notification)
                }
            )
    }
    
    private func setupRefreshTimer() {
        refreshTimer = Timer.scheduledTimer(withTimeInterval: 30.0, repeats: true) { _ in
            self.refreshNotifications()
        }
    }
    
    private func handleNewNotification(_ notification: xyz.sonet.app.grpc.proto.Notification) {
        let newNotification = NotificationItem(from: notification)
        
        // Add to beginning of list
        notifications.insert(newNotification, at: 0)
        
        // Apply current filter
        applyFilter()
        
        // Update unread count
        updateUnreadCount()
        
        // Show local notification if app is in background
        if UIApplication.shared.applicationState != .active {
            showLocalNotification(for: newNotification)
        }
    }
    
    private func applyFilter() {
        switch selectedFilter {
        case .all:
            filteredNotifications = notifications
        case .mentions:
            filteredNotifications = notifications.filter { $0.type == .mention }
        case .likes:
            filteredNotifications = notifications.filter { $0.type == .like }
        case .reposts:
            filteredNotifications = notifications.filter { $0.type == .repost }
        case .follows:
            filteredNotifications = notifications.filter { $0.type == .follow }
        case .replies:
            filteredNotifications = notifications.filter { $0.type == .reply }
        case .hashtags:
            filteredNotifications = notifications.filter { $0.type == .hashtag }
        case .appUpdates:
            filteredNotifications = notifications.filter { $0.type == .appUpdate }
        }
    }
    
    private func updateUnreadCount() {
        unreadCount = notifications.filter { !$0.isRead }.count
    }
    
    private func loadNotificationPreferences() {
        // Load from UserDefaults
        if let data = UserDefaults.standard.data(forKey: "NotificationPreferences"),
           let preferences = try? JSONDecoder().decode(NotificationPreferences.self, from: data) {
            notificationPreferences = preferences
        }
    }
    
    private func requestNotificationPermissions() {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, error in
            if granted {
                DispatchQueue.main.async {
                    UIApplication.shared.registerForRemoteNotifications()
                }
            }
        }
    }
    
    private func showLocalNotification(for notification: NotificationItem) {
        let content = UNMutableNotificationContent()
        content.title = notification.title
        content.body = notification.message
        content.sound = .default
        
        let request = UNNotificationRequest(
            identifier: notification.id,
            content: content,
            trigger: nil
        )
        
        UNUserNotificationCenter.current().add(request)
    }
    
    deinit {
        notificationStream?.cancel()
        refreshTimer?.invalidate()
    }
}

// MARK: - Data Models
struct NotificationItem: Identifiable {
    let id: String
    let type: NotificationType
    let title: String
    let message: String
    let timestamp: Date
    let isRead: Bool
    let sender: UserProfile?
    let targetNote: Note?
    let targetUser: UserProfile?
    let metadata: [String: String]
    
    init(from grpcNotification: xyz.sonet.app.grpc.proto.Notification) {
        self.id = grpcNotification.notificationId
        self.type = NotificationType(from: grpcNotification.type)
        self.title = grpcNotification.title
        self.message = grpcNotification.message
        self.timestamp = Date(timeIntervalSince1970: TimeInterval(grpcNotification.timestamp.seconds))
        self.isRead = grpcNotification.isRead
        self.sender = grpcNotification.hasSender ? UserProfile(from: grpcNotification.sender) : nil
        self.targetNote = grpcNotification.hasTargetNote ? Note(from: grpcNotification.targetNote) : nil
        self.targetUser = grpcNotification.hasTargetUser ? UserProfile(from: grpcNotification.targetUser) : nil
        self.metadata = Dictionary(grpcNotification.metadataMap.mapKeys { $0.key }, uniquingKeysWith: { first, _ in first })
    }
}

enum NotificationType: String, CaseIterable {
    case like = "like"
    case reply = "reply"
    case repost = "repost"
    case follow = "follow"
    case mention = "mention"
    case hashtag = "hashtag"
    case appUpdate = "app_update"
    case system = "system"
    
    init(from grpcType: xyz.sonet.app.grpc.proto.NotificationType) {
        switch grpcType {
        case .NOTIFICATION_TYPE_LIKE: self = .like
        case .NOTIFICATION_TYPE_REPLY: self = .reply
        case .NOTIFICATION_TYPE_REPOST: self = .repost
        case .NOTIFICATION_TYPE_FOLLOW: self = .follow
        case .NOTIFICATION_TYPE_MENTION: self = .mention
        case .NOTIFICATION_TYPE_HASHTAG: self = .hashtag
        case .NOTIFICATION_TYPE_APP_UPDATE: self = .appUpdate
        case .NOTIFICATION_TYPE_SYSTEM: self = .system
        default: self = .system
        }
    }
    
    var icon: String {
        switch self {
        case .like: return "heart.fill"
        case .reply: return "bubble.left.fill"
        case .repost: return "arrow.2.squarepath.fill"
        case .follow: return "person.badge.plus.fill"
        case .mention: return "at"
        case .hashtag: return "number"
        case .appUpdate: return "app.badge.fill"
        case .system: return "bell.fill"
        }
    }
    
    var color: Color { .primary }
}

struct AppUpdateInfo: Identifiable {
    let id: String
    let version: String
    let title: String
    let description: String
    let changelog: [String]
    let isRequired: Bool
    let downloadUrl: String
    let releaseDate: Date
    
    init(from grpcUpdate: xyz.sonet.app.grpc.proto.AppUpdate) {
        self.id = grpcUpdate.updateId
        self.version = grpcUpdate.version
        self.title = grpcUpdate.title
        self.description = grpcUpdate.description
        self.changelog = grpcUpdate.changelogList
        self.isRequired = grpcUpdate.isRequired
        self.downloadUrl = grpcUpdate.downloadUrl
        self.releaseDate = Date(timeIntervalSince1970: TimeInterval(grpcUpdate.releaseDate.seconds))
    }
}

struct NotificationPreferences: Codable {
    var pushEnabled: Bool = true
    var emailEnabled: Bool = true
    var likeNotifications: Bool = true
    var replyNotifications: Bool = true
    var repostNotifications: Bool = true
    var followNotifications: Bool = true
    var mentionNotifications: Bool = true
    var hashtagNotifications: Bool = true
    var appUpdateNotifications: Bool = true
    var systemNotifications: Bool = true
    var quietHoursEnabled: Bool = false
    var quietHoursStart: Date = Calendar.current.date(from: DateComponents(hour: 22, minute: 0)) ?? Date()
    var quietHoursEnd: Date = Calendar.current.date(from: DateComponents(hour: 8, minute: 0)) ?? Date()
    
    func toGRPC() -> xyz.sonet.app.grpc.proto.NotificationPreferences {
        var preferences = xyz.sonet.app.grpc.proto.NotificationPreferences()
        preferences.pushEnabled = pushEnabled
        preferences.emailEnabled = emailEnabled
        preferences.likeNotifications = likeNotifications
        preferences.replyNotifications = replyNotifications
        preferences.repostNotifications = repostNotifications
        preferences.followNotifications = followNotifications
        preferences.mentionNotifications = mentionNotifications
        preferences.hashtagNotifications = hashtagNotifications
        preferences.appUpdateNotifications = appUpdateNotifications
        preferences.systemNotifications = systemNotifications
        preferences.quietHoursEnabled = quietHoursEnabled
        
        var startTimestamp = Timestamp()
        startTimestamp.seconds = Int64(quietHoursStart.timeIntervalSince1970)
        preferences.quietHoursStart = startTimestamp
        
        var endTimestamp = Timestamp()
        endTimestamp.seconds = Int64(quietHoursEnd.timeIntervalSince1970)
        preferences.quietHoursEnd = endTimestamp
        
        return preferences
    }
}